File: algorithms/python/100-SameTree/main.py
#
# https://leetcode.com/problems/same-tree/
# Easy
# Topics: Tree, Depth First Search, Breadth First Search, Binary Tree
#
# This can be solved recursively or iteratively.
#
# The recursive approach is the most intuitive. Trivial checks up front followed by recursive calls for the children
#
# Time Complexity: O(n)
# Space Complexity: O(n)
#
# For the iterative approach, we can use Breadth First Search. In this case we can use a queue. We will use two queues
#
# Time Complexity: O(n)
# Space Complexity: O(n)
#

from typing import List
from typing import Optional
from common.Tree import Tree
from common.TreeNode import TreeNode
import unittest

RECURSIVE = False

if RECURSIVE:
    class Solution:
        def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
            # Trivial Checks
            if p is None and q is None:
                return True
            if p is None or q is None:
                return False

            if p.val != q.val:
                return False

            return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)

else:
    class Solution:
        def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
            queue0: List[TreeNode] = []
            queue1: List[TreeNode] = []

            if p is None and q is None:
                return True
            if p is None or q is None:
                return False

            queue0.append(p)
            queue1.append(q)

            while len(queue0):
                node0 = queue0.pop(0)
                node1 = queue1.pop(0)

                if node0 and node1:
                    if node0.val != node1.val:
                        return False
                    left0 = node0.left
                    left1 = node1.left
                    if left0 and left1:
                        queue0.append(left0)
                        queue1.append(left1)
                    elif not (not left0 and not left1):
                        return False
                    right0 = node0.right
                    right1 = node1.right
                    if right0 and right1:
                        queue0.append(right0)
                        queue1.append(right1)
                    elif not (not right0 and not right1):
                        return False
                elif not (not node0 and not node1):
                    return False
            return True

class TestSolution(unittest.TestCase):
    def setUp(self):
        self.soln = Solution()

    def test_case0(self):
        p0 = Tree([1, 2, 3])
        q0 = Tree([1, 2, 3])
        self.assertTrue(self.soln.isSameTree(p0.root, q0.root))

    def test_case1(self):
        p1 = Tree([1, 2])
        q1 = Tree([1, None, 2])
        self.assertFalse(self.soln.isSameTree(p1.root, q1.root))

    def test_case2(self):
        p2 = Tree([1, 2, 1])
        q2 = Tree([1, 1, 2])
        self.assertFalse(self.soln.isSameTree(p2.root, q2.root))

File: algorithms/cpp/100-SameTree/main.cpp
//
// Created by Benjamin Lee on 1/9/23.
// https://leetcode.com/problems/same-tree/
// Easy
// Topics
//

/*
 * We can approach this iteratively or recusively. Recursively is very easy to understand. We do trivial rejection first,
 * and then recursively to the children.
 *
 * Recursive solution will be:
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 *
 * Iteratively, this is suited for breadth first search. This uses a queue. We will use two queues, one for p and one
 * for q.
 *
 * Time Complexity: O(n)
 * Space Complexity: O(n)
 */


#define CATCH_CONFIG_MAIN
#include <catch2/catch_all.hpp>

#include "../include/BinarySearchTree.h"
#include <queue>

class Solution {
public:
#ifdef RECURSIVE
    bool isSameTree(TreeNode *p, TreeNode *q) {
        // Trivial checks
        if (!p and !q) return true;
        if (!p or !q) return false;

        if (p->val != q->val) return false;

        return isSameTree(p->left, q->left) and isSameTree(p->right, q->right);
    }
#else
    bool isSameTree(TreeNode *p, TreeNode *q) {
        std::queue<TreeNode *> queue0;
        std::queue<TreeNode *> queue1;

        queue0.push(p);
        queue1.push(q);

        while(queue0.size()) {
            auto node0 = queue0.front();
            auto node1 = queue1.front();

            queue0.pop();
            queue1.pop();

            // Trivial checks
            if (node0 and node1) {
                if (node0->val != node1->val) return false;

                auto left0 = node0->left;
                auto left1 = node1->left;
                if (left0 and left1) {
                    queue0.push(node0->left);
                    queue1.push(node1->left);
                } else if (!(!left0 and !left1)) {
                    return false;
                }

                auto right0 = node0->right;
                auto right1 = node1->right;
                if (right0 and right1) {
                    queue0.push(node0->right);
                    queue1.push(node1->right);
                } else if (!(!right0 and !right1)) {
                    return false;
                }
            } else if (!(!node0 and !node1)) {
                return false;
            }
        }

        return true;
    }
#endif /* RECURSIVE */
};

TEST_CASE("SameTree", "algorithm") {
    Solution soln;
    Tree p0{1, 2, 3};
    Tree q0{1, 2, 3};
    Tree p1{1, 2};
    Tree q1{1, std::numeric_limits<int>::min(), 2};
    Tree p2{1, 2, 1};
    Tree q2{1, 1, 2};

    CHECK(soln.isSameTree(p0.root, q0.root) == true);
    CHECK(soln.isSameTree(p1.root, q1.root) == false);
    CHECK(soln.isSameTree(p2.root, q2.root) == false);
}